Core/Src/internalResistance.c

```diff
@@ -0,0 +1,221 @@
+        if(currentAvgBuffer[i] > -900.0f)
+        {
+            if(currentAvgBuffer[i] > maxCurrent)
+            {
+                maxCurrentIndex = i;
+            }
+            else if(currentAvgBuffer[i] < minCurrent)
+            {
+                minCurrentIndex = i;
+            }
+        }
+    }
+
+    //  Calculate greatest dI and dV contained in buffer
+    /*  If avg buffer saturated with -1000.0f placeholder values, default index 0 will still be set, 
+        and delta current will equal 0 (-1000.0f - -1000.0f)    */
+    float deltaCurrent = currentAvgBuffer[maxCurrentIndex] - currentAvgBuffer[minCurrentIndex];
+
+    // If delta current threshold is met, internal resistance can be calulated and updated
+    // This threshold should always exceed 0, so if avg buffer is saturated with -1000.0f, operation will be skipped
+    if(fabs(deltaCurrent) >= IR_CALC_MIN_CURRENT_DELTA)
+    {
+        // Cycle through every BMB and every brick in the accumulator
+        for(int32_t i = 0; i < NUM_BMBS_PER_PACK; i++)
+        {
+            for(int32_t j = 0; j < NUM_BRICKS_PER_BMB; j++)
+            {
+                // If average current buffer containes placeholder value, -1000.0f, it is excluded from this calculation cycle
+                if((voltageAvgBuffer[i][j][maxCurrentIndex] > -900.0f) && (voltageAvgBuffer[i][j][minCurrentIndex] > -900.0f))
+                {
+                    float deltaVoltage = voltageAvgBuffer[i][j][maxCurrentIndex] - voltageAvgBuffer[i][j][minCurrentIndex];
+                    bms->bmb[i].brickResistance[j] = deltaVoltage / deltaCurrent;
+                }
+            }    
+        }
+    }
+
+    // Static data ready vars set to 0 for next discrete buffer filling cycle 
+    currentDataReady = false;
+    voltageDataReady = false;
+
+    // Reset discrete buffer pointers so data begins filling from index 0 
+    discreteCurrentBufferIndex = 0;
+    discreteVoltageBufferIndex = 0;
+}
+
+/* ==================================================================== */
+/* =================== GLOBAL FUNCTION DEFINITIONS ==================== */
+/* ==================================================================== */
+
+void putCurrentBuffer(Bms_S* bms)
+{   
+    // If any data in the discrete buffer is from a faulty sensor, set that the current data is bad
+    if(bms->currentSensorStatusLO == SENSE_GOOD)
+    {
+        // Update discrete data buffers with current bms current data
+        currentDiscreteBuffer[discreteCurrentBufferIndex] = (bms->tractiveSystemCurrent / 25.0f);
+    }
+    else
+    {
+        currentDataGood = false;
+    }
+
+    // Cycle buffer and wrap to 0 if needed
+    if(++discreteCurrentBufferIndex >= DISCRETE_CURRENT_BUFFER_SIZE)
+    {
+        discreteCurrentBufferIndex = 0;
+
+        // Indicate that the discrete current buffer has been filled
+        currentDataReady = true;
```

Comment:
I'm not a fan of this `currentDataReady` and `voltageDataReady`. Is there a better way of handling this? Ideally the two should never be able to get out of synch so that you can guarantee that all the data is time correlated properly

Reply:
We are filling the discrete buffers at different frequencies, and from different functions- So doesn't this current implementation guarantee that we add to the average buffer only as soon as both buffers fill? This also means if data does get out of synch (by even instruction cycles), it stays time correlated: If one buffer fills, it will keep filling in a circular manner until the other fills. Then the data from both discrete buffers, meaning the most recent data that has filled both of them, is added to the average buffer.

Reply:
I see what you are saying. Is it worth it to have these extra samples for the current? It might be simplest to just have a function that updates both voltages and currents. How much are we benefiting from 50 samples vs 10 samples? Also can we just use the calculated tractiveSystemCurrent value? We can have a sensor status for that value and just check that.

Reply:
This would also allow us to get rid of voltageDataReady and just see if `bufferIdx == bufferSize` and if so call the averaging function